stan/math/rev/fun/inverse.hpp

```diff
@@ -75,16 +22,22 @@ class inverse_vari : public vari {
  * size zero).
  * @throw std::invalid_argument if the matrix is not square.
  */
-inline matrix_v inverse(const matrix_v &m) {
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline plain_type_t<T> inverse(const T& m) {
   check_square("inverse", "m", m);
   if (m.size() == 0) {
-    return {};
+    return m;
   }
 
-  matrix_v res(m.rows(), m.cols());
-  internal::inverse_vari *baseVari = new internal::inverse_vari(m);
-  res.vi() = Eigen::Map<matrix_vi>(baseVari->vari_ref_A_inv_, res.rows(),
-                                   res.cols());
+  arena_t<plain_type_t<T>> arena_m = m;
+  arena_t<Eigen::MatrixXd> res_val = value_of(arena_m).inverse();
+  arena_t<plain_type_t<T>> res = res_val;
+
+  reverse_pass_callback([res, res_val, arena_m]() mutable {
+    Eigen::MatrixXd res_adj = res.adj();
+    arena_m.adj() -= res_val.transpose() * res_adj * res_val.transpose();
```

Comment:
```suggestion
    arena_m.adj() -= res_val.transpose() * res.adj() * res_val.transpose();
```

Reply:
Whenever I try to do this I get errors in the matrix multiplication that come back to:
```
./stan/math/rev/fun/inverse.hpp:38:21: note: in instantiation of function template specialization
      'Eigen::MatrixBase<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double>, -1, -1, 0, -1, -1>, 8, Eigen::Stride<0, 0> >
      >::adj_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double>, -1, -1, 0, -1, -1>, 8, Eigen::Stride<0, 0> > >
      >::operator-=<Eigen::Product<Eigen::Product<Eigen::Transpose<Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 8, Eigen::Stride<0, 0> > >,
      Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double>, -1, -1, 0, -1, -1>, 8, Eigen::Stride<0, 0> > >::adj_Op,
      Eigen::Map<Eigen::Matrix<stan::math::var_value<double>, -1, -1, 0, -1, -1>, 8, Eigen::Stride<0, 0> > >, 0>, Eigen::Transpose<Eigen::Map<Eigen::Matrix<double, -1, -1,
      0, -1, -1>, 8, Eigen::Stride<0, 0> > >, 0> >' requested here
      arena_m.adj() -= res_val.transpose() * res.adj() * res_val.transpose();
```

The error itself is:
```
lib/eigen_3.3.7/Eigen/src/Core/products/GeneralMatrixMatrix.h:228:22: error: no matching member function for call to 'coeffRef'
              &m_rhs.coeffRef(0,col), m_rhs.outerStride()
```

Any ideas?

Reply:
It's the bug where multiplying two ```CWiseUnaryViews``` fails. This was why Steve was introducing the ```adj_op()``` member to always return a ```CWiseUnaryOp``` regardless of const-ness.

In this case you'll either have to eval the ```res.adj()``` or add const-ness to ```res``` before calling the adj member. Maybe something like:
```
const_cast<const decltype(res)>(res).adj()
```
?

Reply:
@andrjohns thanks. I just switched to:

```
arena_m.adj() -= res_val.transpose() * res.adj_op() * res_val.transpose();
```

and it worked. Is `adj_op` supposed to be the same as `adj`?

Reply:
Didn't realise that was in yet, neat! Yeah, so the issue with ```adj``` is that will return a different Eigen expression depending on whether the input matrix is const or not:
```
Eigen::Matrix<var, -1, -1> non_const(R, C);
non_const.adj()  // Returns a CWiseUnaryView, which can be used as either an l-value or an r-value

const Eigen::Matrix<var, -1, -1> const(R, C);
const.adj()  // Returns a CWiseUnaryOp, which can only be used as an r-value
```

But Eigen has a bug with multiplying ```CWiseUnaryViews``` together, so Steve introduced the ```adj_op``` member which will always return a CWiseUnaryOp:
```
Eigen::Matrix<var, -1, -1> non_const(R, C);
non_const.adj_op()  // Returns a CWiseUnaryOp

const Eigen::Matrix<var, -1, -1> const(R, C);
const.adj_op()  // Returns a CWiseUnaryOp
```

Reply:
Righto, from now on if `adj()` gives me compiler errors I'm just gonna switch to `adj_op()`. Thanks!